### PR TITLE
Monitoring & Logging: Add two probes into abacus-rse instead Fix #4261

### LINKED ISSUE #4261: Move deletable and obsolete replica checks from probes into Abacus-RSE
Motivation
----------

These "probes" are really required for correct Rucio functioning and should be in a daemon rather than in probes

Modification
------------

### PR DESCRIPTION
**Pull request by ericvaandering**
This adds the functionality of two probes into abacus-rse. There was discussion that other things could be added?

### CODE REVIEW

File: lib/rucio/core/rse_counter.py [hunk 1]
```diff
@@ -27,7 +30,7 @@ def add_counter(rse_id, session=None):
     :param rse_id:  The id of the RSE.
     :param session: The database session in use.
     """
-    models.RSEUsage(rse_id=rse_id, source='rucio', files=0, used=0).\
+    models.RSEUsage(rse_id=rse_id, source='rucio', files=0, used=0). \
```
Comment (bziemons) [inline lib/rucio/core/rse_counter.py]: if you want to format these lines, please move the dot to the next line and the escaping backslash wont be needed anymore.

File: lib/rucio/core/rse_counter.py [hunk 2]
```diff
@@ -12,11 +12,14 @@
 # - Hannes Hansen, <hannes.jakob.hansen@cern.ch>, 2018-2019
 # - Brandon White, <bjwhite@fnal.gov>, 2019
```
Comment (bziemons) [inline lib/rucio/core/rse_counter.py:15]: please update the copyright either by using `tools/add_header -i <filename>` or manually

File: lib/rucio/core/rse_counter.py [hunk 3]
```diff
@@ -138,6 +142,49 @@ def update_rse_counter(rse_id, session=None):
...
+        obsolete_files = 0
+
+    # Merge (INSERT or UPDATE) the values
+    deletable = models.RSEUsage(rse_id=rse_id, used=deletable_bytes, files=deletable_files, source='expired')
+    session.merge(deletable)
+    obsolete = models.RSEUsage(rse_id=rse_id, used=obsolete_bytes, files=obsolete_files, source='obsolete')
+    session.merge(obsolete)
+
+    # fill_rse_counter_history_table takes care of the history for all types
+    return
```
Comment (bziemons) [inline lib/rucio/core/rse_counter.py]: remove unnecessary return

File: lib/rucio/core/rse_counter.py [hunk 4]
```diff
@@ -147,4 +194,5 @@ def fill_rse_counter_history_table(session=None):
     """
     RSEUsageHistory = models.RSEUsage.__history_mapper__.class_
     for usage in session.query(models.RSEUsage).all():
-        RSEUsageHistory(rse_id=usage['rse_id'], used=usage['used'], files=usage['files'], source=usage['source']).save(session=session)
+        RSEUsageHistory(rse_id=usage['rse_id'], used=usage['used'], files=usage['files'], source=usage['source']).save(
+            session=session)
```
Comment (bziemons) [inline lib/rucio/core/rse_counter.py]: please don't break a line here, this does not make it more readable

File: lib/rucio/daemons/abacus/rse.py [hunk 5]
```diff
@@ -71,15 +72,23 @@ def rse_update(once=False):
...
                         break
                     start_time = time.time()
                     update_rse_counter(rse_id=rse_id)
-                    logging.debug('rse_update[%s/%s]: update of rse "%s" took %f' % (heartbeat['assign_thread'], heartbeat['nr_threads'] - 1, rse_id, time.time() - start_time))
+                    logging.debug('rse_update[%s/%s]: counter update of rse "%s" took %f' %
+                                  (heartbeat['assign_thread'], heartbeat['nr_threads'] - 1, rse_id,
+                                   time.time() - start_time))
+                    start_time = time.time()
+                    update_replica_counts(rse_id=rse_id)
+                    logging.debug('rse_update[%s/%s]: replica update of rse "%s" took %f' %
```
Comment (bziemons) [inline lib/rucio/daemons/abacus/rse.py]: if you're adding new log messages anyway, could you please create a `logger` after fetching the heartbeat in line 65 and use that?

File: lib/rucio/core/rse_counter.py (hunk 2 above)
Comment (bziemons) [reply lib/rucio/core/rse_counter.py:15]: also please update the copyright in the other files

File: lib/rucio/daemons/abacus/rse.py [hunk 6]
```diff
@@ -37,7 +37,8 @@
 from rucio.common.logging import setup_logging
 from rucio.common.utils import get_thread_with_periodic_running_function
 from rucio.core.heartbeat import live, die, sanity_check
-from rucio.core.rse_counter import get_updated_rse_counters, update_rse_counter, fill_rse_counter_history_table
+from rucio.core.rse_counter import (fill_rse_counter_history_table, get_updated_rse_counters,
+                                    update_replica_counts, update_rse_counter)
```
Comment (bziemons) [inline lib/rucio/daemons/abacus/rse.py:42]: I remember there was a discussion about this notation already. Is this just autoformatting?

File: lib/rucio/core/rse_counter.py [hunk 7]
```diff
@@ -138,6 +142,49 @@ def update_rse_counter(rse_id, session=None):
...
+    """
+    none_value = None
+
+    # Get the counds of obsolete and deletably (expired) files
+    obsolete_files, obsolete_bytes = (session
+                                      .query(func.count(), func.sum(models.RSEFileAssociation.bytes))
+                                      .with_hint(models.RSEFileAssociation,
+                                                 "INDEX_FFS(REPLICAS REPLICAS_TOMBSTONE_IDX)", 'oracle')
+                                      .filter(models.RSEFileAssociation.rse_id == rse_id,
+                                              models.RSEFileAssociation.tombstone != none_value,
```
Comment (bziemons) [inline lib/rucio/core/rse_counter.py]: please use `null()` from sqlalchemy instead

File: lib/rucio/core/rse_counter.py [hunk 8]
```diff
@@ -138,6 +142,49 @@ def update_rse_counter(rse_id, session=None):
...
+    Read the replica counts and update rse_usage.
+
+    :param rse_id:   The rse_id to update.
+    :param session:  Database session in use.
+    """
+    none_value = None
+
+    # Get the counds of obsolete and deletably (expired) files
+    obsolete_files, obsolete_bytes = (session
+                                      .query(func.count(), func.sum(models.RSEFileAssociation.bytes))
```
Comment (bziemons) [inline lib/rucio/core/rse_counter.py:153]: is there a way to make these blocks a bit better readable?

**Review by bziemons — CHANGES_REQUESTED** [review]
my comments are mostly cosmetics, functionality-wise I don't see a problem. I am not sure how `RSEUsage` is updated so far. This would override any values set by other means and is not configurable.

File: lib/rucio/core/rse_counter.py (hunk 8 above)
Comment (ericvaandering) [reply lib/rucio/core/rse_counter.py:153]: Not sure what's unreadable about it. Using () to enclose multiple lines is now preferred and it clearly separates the clauses. Indentation is determined by PyCharm

File: lib/rucio/core/rse_counter.py (hunk 1 above)
Comment (ericvaandering) [reply lib/rucio/core/rse_counter.py]: These were all short enough to combine on a single line.

**Comment by ericvaandering** [comment]
I think the only test which didn't work was the merging test. But that should be solved now. Any comments on the general approach of where this is located?

**Comment by bziemons** [comment]
> I think the only test which didn't work was the merging test. But that should be solved now. Any comments on the general approach of where this is located?

you're not following the [branch naming guidelines](https://github.com/rucio/rucio/blob/master/CONTRIBUTING.rst#contributing). Branches prefixed with `feature-` will not be cherry-picked to the latest release branch. The release-patch-setup fails if the cherry-picking fails and master and the latest release branch diversed too much for your changes.

File: lib/rucio/daemons/abacus/rse.py [hunk 9]
```diff
@@ -62,32 +64,43 @@ def rse_update(once=False):
...
             start = time.time()  # NOQA
             rse_ids = get_updated_rse_counters(total_workers=heartbeat['nr_threads'],
                                                worker_number=heartbeat['assign_thread'])
-            logging.debug('Index query time %f size=%d' % (time.time() - start, len(rse_ids)))
+            logger(logging.DEBUG, 'Index query time %f size=%d' % (time.time() - start, len(rse_ids)))
 
             # If the list is empty, sent the worker to sleep
             if not rse_ids and not once:
-                logging.info('rse_update[%s/%s] did not get any work' % (heartbeat['assign_thread'], heartbeat['nr_threads'] - 1))
+                logger(logging.INFO, 'rse_update[%s/%s] did not get any work' %
```
Comment (bziemons) [inline lib/rucio/daemons/abacus/rse.py:79]: double prefix

File: lib/rucio/daemons/abacus/rse.py [hunk 10]
```diff
@@ -62,32 +64,43 @@ def rse_update(once=False):
...
+                       (heartbeat['assign_thread'], heartbeat['nr_threads'] - 1))
                 time.sleep(10)
             else:
                 for rse_id in rse_ids:
                     if graceful_stop.is_set():
                         break
                     start_time = time.time()
                     update_rse_counter(rse_id=rse_id)
-                    logging.debug('rse_update[%s/%s]: update of rse "%s" took %f' % (heartbeat['assign_thread'], heartbeat['nr_threads'] - 1, rse_id, time.time() - start_time))
+                    logger(logging.DEBUG, 'rse_update[%s/%s]: counter update of rse "%s" took %f' %
```
Comment (bziemons) [inline lib/rucio/daemons/abacus/rse.py:88]: double prefix

File: lib/rucio/daemons/abacus/rse.py [hunk 11]
```diff
@@ -62,32 +64,43 @@ def rse_update(once=False):
...
+                           (heartbeat['assign_thread'], heartbeat['nr_threads'] - 1, rse_id,
+                            time.time() - start_time))
+                    start_time = time.time()
+                    update_replica_counts(rse_id=rse_id)
+                    logger(logging.DEBUG, 'rse_update[%s/%s]: replica update of rse "%s" took %f' %
```
Comment (bziemons) [inline lib/rucio/daemons/abacus/rse.py:93]: double prefix

**Review by bziemons — CHANGES_REQUESTED** [review]

File: lib/rucio/core/rse_counter.py (hunk 8 above)
Comment (bziemons) [reply lib/rucio/core/rse_counter.py:153]: How about this:
```diff
+    obsolete_files, obsolete_bytes = (
+        session.query(func.count(), func.sum(models.RSEFileAssociation.bytes))
+               .with_hint(models.RSEFileAssociation, "INDEX_FFS(REPLICAS REPLICAS_TOMBSTONE_IDX)", 'oracle')
+               .filter(models.RSEFileAssociation.rse_id == rse_id,
+                       models.RSEFileAssociation.tombstone != null(),
+                       models.RSEFileAssociation.tombstone == OBSOLETE)
+               .one()
+    )
```

Comment (bari12) [reply lib/rucio/core/rse_counter.py:153]: 👍 I prefer that one too. Both are PEP8 compatible I guess, but for me the lonely `session` looks strange. Bens version has the advantage of saving line-length space as well.

**Comment by ericvaandering** [comment]
Ok I have taken the code (after resolving comments) and put it into https://github.com/rucio/rucio/pull/4437 with the correct branch name.

**Pull request closed without merging**